Assets/Weapons/FireWeapon.cs

```diff
@@ -0,0 +1,334 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class FireWeapon : Weapon
+{
+
+    [Header("Fire Weapon")] 
+    public FireWeaponData fireWeaponData;
+
+    //Gun muzzle wil be used for FX
+    public Transform gunMuzzle;
+    //The transform where the ray casts will be emitted
+    public Transform dispersionTransform;
+
+    private int _currentAmmo;
+    private bool _isReloading;
+
+    //Fire rate global variables
+    private float _fireRateTimer;
+    private float _fireRateCurveTimer;
+
+    //@TODO: Fix initial spamming
+    private bool _wantsToFire;
+    private bool _canFire = true;
+
+    //Dispersion global variables
+    private float _currentDispersion;
+    private float _dispersionCurveTimer;
+
+    
+#if UNITY_EDITOR
+    [Header("Debug")]
+    [SerializeField] private bool log;
+    [SerializeField] private bool drawGizmos;
+    [SerializeField] private float gizmosDuration = .5f;
+#endif
+    
+    public override int UsesLeft()
+    {
+        return _currentAmmo;
+    }
+    
+    /// <summary>
+    /// Use Functionality
+    /// </summary>
+    public override void Use(bool pressed)
```

Comment:
You can make a performed and exited function (two separate ones) to avoid having the pressed bool